codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/EndpointResolverMiddleware.kt

```diff
@@ -5,34 +5,98 @@
-import software.amazon.smithy.model.Model
-import software.amazon.smithy.model.shapes.OperationShape
+import software.amazon.smithy.aws.swift.codegen.AWSServiceTypes
+import software.amazon.smithy.aws.swift.codegen.AWSSwiftDependency
+import software.amazon.smithy.aws.swift.codegen.ENDPOINT_PARAMS
+import software.amazon.smithy.aws.swift.codegen.ENDPOINT_RESOLVER
+import software.amazon.smithy.codegen.core.Symbol
+import software.amazon.smithy.swift.codegen.ClientRuntimeTypes
+import software.amazon.smithy.swift.codegen.Middleware
+import software.amazon.smithy.swift.codegen.SwiftDependency
 import software.amazon.smithy.swift.codegen.SwiftWriter
-import software.amazon.smithy.swift.codegen.integration.middlewares.handlers.MiddlewareShapeUtils
-import software.amazon.smithy.swift.codegen.middleware.MiddlewarePosition
-import software.amazon.smithy.swift.codegen.middleware.MiddlewareRenderable
-import software.amazon.smithy.swift.codegen.middleware.MiddlewareStep
+import software.amazon.smithy.swift.codegen.integration.steps.OperationBuildStep
 
+/**
+ * Generates endpoint middleware for the service.
+ */
 class EndpointResolverMiddleware(
-    val model: Model,
-    val symbolProvider: SymbolProvider
-) : MiddlewareRenderable {
+    private val writer: SwiftWriter,
+    inputSymbol: Symbol,
+    outputSymbol: Symbol,
+    outputErrorSymbol: Symbol
+) : Middleware(writer, inputSymbol, OperationBuildStep(outputSymbol, outputErrorSymbol)) {
+
+    override val id: String = "EndpointResolverMiddleware"
+
+    override val typeName =
+        "EndpointResolverMiddleware<$outputSymbol: ${ClientRuntimeTypes.Http.HttpResponseBinding}, $outputErrorSymbol: ${ClientRuntimeTypes.Http.HttpResponseBinding}>"
+
+    override val properties: MutableMap<String, Symbol> = mutableMapOf(
+        ENDPOINT_RESOLVER to AWSServiceTypes.EndpointResolver, ENDPOINT_PARAMS to AWSServiceTypes.EndpointParams
+    )
+
+    override fun generateInit() {
+        writer.openBlock(
+            "public init($ENDPOINT_RESOLVER: \$N, $ENDPOINT_PARAMS: \$N) {",
+            "}",
+            AWSServiceTypes.EndpointResolver,
+            AWSServiceTypes.EndpointParams
+        ) {
+            writer.write("self.\$L = \$L", ENDPOINT_RESOLVER, ENDPOINT_RESOLVER)
+            writer.write("self.\$L = \$L", ENDPOINT_PARAMS, ENDPOINT_PARAMS)
+        }
+    }
+
+    override fun generateMiddlewareClosure() {
+        writer.addImport(SwiftDependency.CLIENT_RUNTIME.packageName)
+        writer.addImport(AWSSwiftDependency.AWS_CLIENT_RUNTIME.packageName)
+        writer.write("let endpoint = try endpointResolver.resolve(params: endpointParams)")
+
+        writer.write("")
+        writer.openBlock("guard let authScheme = endpoint.authScheme(name: \"sigv4\") else {", "}") {
+            writer.write(
+                "throw \$N.client(ClientError.unknownError((\"Unable to resolve endpoint. Missing auth scheme.\")))",
+                errorType
+            )
```

Comment:
**Question**: Is the error here always a _missing_ auth scheme? Or could it also be an _unsupported_ scheme?